### PR TITLE
Fix cocurrent upload package

### PR DESCRIPTION
**Pull request by lunny**
When inserting failed, then check if the record does exist, if yes, then just return the record duplicated.

### CODE REVIEW

File: models/packages/package.go [hunk 1]
```diff
@@ -128,14 +128,21 @@ func TryInsertPackage(ctx context.Context, p *Package) (*Package, error) {
...
 	}
 	if _, err = e.Insert(p); err != nil {
+		has, err := e.Exist(key)
+		if err != nil {
+			return nil, err
+		}
+		if has {
+			return key, ErrDuplicatePackage
+		}
 		return nil, err
```
Comment (wxiaoguang) [inline models/packages/package.go:146]: That's wrong. You might get `has=false, err=nil` then `return nil, nil`

File: models/packages/package.go [hunk 2]
```diff
@@ -128,14 +128,21 @@ func TryInsertPackage(ctx context.Context, p *Package) (*Package, error) {
 		LowerName: p.LowerName,
 	}
 
-	has, err := e.Get(key)
+	has, err := e.Exist(key)
```
Comment (KN4CK3R) [inline models/packages/package.go:131]: I think this is wrong because `Exist` does not populate the object with the content from the database.

**Comment by KN4CK3R** [comment]
And this doesn't fix the current problem because the error does not occur in these methods but when the transaction is committed.

**Comment by wxiaoguang** [comment]
> And this doesn't fix the current problem because the error does not occur in these methods but when the transaction is committed.

Although I haven't looked into the problem, I think using transaction correctly will resolve the concurrency problem.


Database transactions will acquire `write` lock for all DML (insert/update/...) statements

**Comment by wxiaoguang** [comment]
Will be replaced by 
* https://github.com/go-gitea/gitea/pull/21862
* https://github.com/KN4CK3R/gitea/pull/3

**Pull request closed without merging**